src/x86/intel_joule_expansion.c

```diff
@@ -149,6 +170,114 @@ mraa_joule_expansion_board()
     b->spi_bus[4].bus_id = 32765;
     b->spi_bus[4].slave_s = 2;
 
+    b->uart_dev_count = 2;
+    b->def_uart_dev = 0;
+    b->uart_dev[0].device_path = "/dev/ttyS0";
+    b->uart_dev[0].rx = 68;
+    b->uart_dev[0].tx = 7;
+    b->uart_dev[1].device_path = "/dev/ttyS1";
+    b->uart_dev[1].rx = 24;
+    b->uart_dev[1].tx = 22;
+
+    // Aio shield detection
+    // Only if i2c_designware.0 was successfully found
+    if (i2c_aio_bus != -1) {
+
+        syslog(LOG_NOTICE, "Attempting shield autodetection on I2C bus %d...", i2c_aio_bus);
+        mraa_i2c_context i2c = mraa_i2c_init_raw(i2c_aio_bus);
+
+        if (i2c == NULL)
+            syslog(LOG_ERR, "Failed to open I2C bus: %d", i2c_aio_bus);
+        else {
+            if (mraa_i2c_address(i2c, 0x49) != MRAA_SUCCESS)
+                syslog(LOG_ERR, "Failed to set I2C address: %d", 0x49);
+            if (mraa_i2c_read_word_data(i2c, 0x01) < 0) {
+                syslog(LOG_NOTICE, "No device at I2C address 0x49");
+
+                // no DFRobot shield, try Grove
+                if (mraa_i2c_address(i2c, 0x48) != MRAA_SUCCESS)
+                    syslog(LOG_ERR, "Failed to set I2C address: %d", 0x48);
+                if (mraa_i2c_read_word_data(i2c, 0x01) < 0)
+                    syslog(LOG_NOTICE, "No device at I2C address 0x48");
+                    // no shield
+
+                else {
+                    // load ads1015 at 0x48
+                    syslog(LOG_NOTICE, "Loading ti-ads1015 module for Grove Shield");
+                    if (system("modprobe ti-ads1015") < 0) {
```

Comment:
Hmm.

Well, as a first observation, **IF** you want to do this kind of thing I'd recommend 2 things:
1. Use fork/exec rather than system().  Since all you want to do is call an executable, there is no point in invoking an intermediate shell.  It's more complicated, particularly in detecting an error (you'll need a sig handler), but will be faster and safer.  Since it's not done very often though, this is probably overkill.
2. Regardless of which method you use to spawn the process, **ALWAYS** use a complete path to the executable you are spawning -- in this case **_/sbin/modprobe_**.  This avoids a whole host of "Hacking 101" possibilities via PATH or other environment manipulation.

As for whether it should be done at all in this case - I am definitely a fan of "It should Just Work®".  But in this case the functionality depends on external kernel modules being loaded (or even available), so I think I am with @arfoll and @alext-mkrs here:  This should be done at system startup or manually by the user (documentation should make this clear).  MRAA should expect everything it needs from the kernel/OS to be there and be ready.

In my view, it would make sense to just have all of these potential drivers be loaded at boot (via modules.conf or whatever), _assuming_ of course that MRAA can reliably detect at runtime whether these devices are in fact available (hardware present/driver loaded), and use them if they are.